spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteDataFilesSparkAction.java

```diff
@@ -0,0 +1,429 @@
+
+    ExecutorService rewriteService = Executors.newFixedThreadPool(maxConcurrentFileGroupActions,
+        new ThreadFactoryBuilder().setNameFormat("Rewrite-Service-%d").build());
+
+    ExecutorService committerService = Executors.newSingleThreadExecutor(
+        new ThreadFactoryBuilder().setNameFormat("Committer-Service").build());
+
+    int groupsPerCommit = partialProgressEnabled ?
+        IntMath.divide(totalGroups, maxCommits, RoundingMode.CEILING) :
+        totalGroups;
+
+    AtomicBoolean stillRewriting = new AtomicBoolean(true);
+    ConcurrentHashMap<String, Pair<FileGroupInfo, FileGroupRewriteResult>> results = new ConcurrentHashMap<>();
+    ConcurrentLinkedQueue<String> completedRewriteIds = new ConcurrentLinkedQueue<>();
+    ConcurrentLinkedQueue<String> completedCommitIds = new ConcurrentLinkedQueue<>();
+
+    // Partial progress commit service
+    committerService.execute(() -> {
+      while (stillRewriting.get() || completedRewriteIds.size() > 0) {
+        Thread.yield();
+        // Either we have a full commit group, or we have completed writing and need to commit what is left over
+        if (completedRewriteIds.size() > groupsPerCommit ||
+            (!stillRewriting.get() && completedRewriteIds.size() > 0)) {
+
+          Set<String> batch = Sets.newHashSetWithExpectedSize(groupsPerCommit);
+          for (int i = 0; i < groupsPerCommit && !completedRewriteIds.isEmpty(); i++) {
+            batch.add(completedRewriteIds.poll());
+          }
+
+          try {
+            commitFileGroups(batch);
+            completedCommitIds.addAll(batch);
+          } catch (Exception e) {
+            LOG.error("Failure during rewrite commit process, partial progress enabled. Ignoring", e);
+          }
+        }
+      }
+    });
+
+    // Start rewrite tasks
+    Tasks.foreach(jobStream.iterator())
+        .suppressFailureWhenFinished()
+        .executeWith(rewriteService)
+        .noRetry()
+        .onFailure((info, exception) -> {
+          LOG.error("Failure during rewrite process for group {}", info.first(), exception);
+          abortFileGroup(info.first().groupID);
+        })
+        .run(infoListPair ->
+            rewriteFiles(infoListPair, totalGroups, numGroupsPerPartition, strategy, completedRewriteIds, results));
+
+    stillRewriting.set(false);
+    committerService.shutdown();
+
+    try {
+      committerService.awaitTermination(10, TimeUnit.MINUTES);
+    } catch (InterruptedException e) {
+      throw new RuntimeException("Cannot complete commit for rewrite, commit service interuptted", e);
+    }
+
+    return new Result(results.values().stream().collect(Collectors.toMap(Pair::first, Pair::second)));
+  }
+
+  @Override
+  public Result execute() {
+    RewriteStrategy strategy =  rewriteStrategy(strategyType).options(this.options());
+    validateOptions(strategy);
+
+    CloseableIterable<FileScanTask> files = files();
```

Comment:
I think it would be simpler if you grouped the logic in `files()`, `filterAndGroupFiles()` and closing of resources into one method. You could call it something like `planFileGroups`.

```
  private Map<StructLike, List<List<FileScanTask>>> planFileGroups(RewriteStrategy strategy) {
    CloseableIterable<FileScanTask> fileScanTasks = table.newScan()
        .filter(filter)
        .ignoreResiduals()
        .planFiles();

    try {
      Map<StructLike, List<FileScanTask>> filesByPartition = Streams.stream(fileScanTasks)
          .collect(Collectors.groupingBy(task -> task.file().partition()));

      return filesByPartition.entrySet().stream()
          .collect(Collectors.toMap(Map.Entry::getKey, entry -> toFileGroups(entry.getValue(), strategy)));

    } finally {
      try {
        fileScanTasks.close();
      } catch (IOException io) {
        LOG.error("Cannot properly close file iterable while planning for rewrite", io);
      }
    }
  }

```

I'd also try to keep collectors as simple as possible by putting the logic into a separate method. For example, we can have something like this:

```
  private List<List<FileScanTask>> toFileGroups(List<FileScanTask> tasks, RewriteStrategy strategy) {
    Iterable<FileScanTask> filtered = strategy.selectFilesToRewrite(tasks);
    Iterable<List<FileScanTask>> groupedTasks = strategy.planFileGroups(filtered);
    return ImmutableList.copyOf(groupedTasks);
  }
```

Reply:
Then it would be just a single line in `execute` that would call `planFileGroups`.

Reply:
I think we currently don't remove partitions that end up getting 0 file groups after filtering.

Reply:
Maybe, we should replace the second stream API part?

```
  private Map<StructLike, List<List<FileScanTask>>> planFileGroups(RewriteStrategy strategy) {
    CloseableIterable<FileScanTask> fileScanTasks = table.newScan()
        .filter(filter)
        .ignoreResiduals()
        .planFiles();

    try {
      Map<StructLike, List<FileScanTask>> filesByPartition = Streams.stream(fileScanTasks)
          .collect(Collectors.groupingBy(task -> task.file().partition()));

      Map<StructLike, List<List<FileScanTask>>> fileGroupsByPartition = Maps.newHashMap();

      filesByPartition.forEach((partition, tasks) -> {
        List<List<FileScanTask>> fileGroups = toFileGroups(tasks, strategy);
        if (fileGroups.size() > 0) {
          fileGroupsByPartition.put(partition, fileGroups);
        }
      });

      return fileGroupsByPartition;
    } finally {
      try {
        fileScanTasks.close();
      } catch (IOException io) {
        LOG.error("Cannot properly close file iterable while planning for rewrite", io);
      }
    }
  }
```

Reply:
I think this is a good change, implementing